src/content/learn/typescript.md

```diff
@@ -0,0 +1,433 @@
+    default:
+      throw new Error("Unknown action");
+  }
+}
+
+export default function App() {
+  const [state, dispatch] = useReducer(stateReducer, initialState);
+
+  const addFive = () => dispatch({ type: "setCount", value: state.count + 5 });
+  const reset = () => dispatch({ type: "reset" });
+
+  return (
+    <div>
+      <h1>Welcome to my counter</h1>
+
+      <p>Count: {state.count}</p>
+      <button onClick={addFive}>Add 5</button>
+      <button onClick={reset}>Reset</button>
+    </div>
+  );
+}
+
+```
+
+```js App.js hidden
+import AppTSX from "./App.tsx";
+export default App = AppTSX;
+```
+
+</Sandpack>
+
+
+We are using TypeScript in a few key places:
+
+ - `interface State` describes the shape of the reducer's state.
+ - `type CounterAction` describes the different actions which can be dispatched to the reducer.
+ - `const initialState: State` provides a type for the initial state, and also the type which is used by `useReducer` by default.
+ - `stateReducer(state: State, action: CounterAction): State` sets the types for the reducer function's arguments and return value.
+
+A more explicit alternative to setting the type on `initialState` is to provide a type argument to `useReducer`:
+
+```ts
+const initialState = { count: 0 };
+
+export default function App() {
+  const [state, dispatch] = useReducer<State>(stateReducer, initialState);
+}
+```
+
+### `useContext` {/*typing-usecontext*/}
+
+The [`useContext` hook](/reference/react/useContext) is a technique for passing data down the component tree without having to pass props through components. It is used by creating a provider component and often by creating a hook to consume the value in a child component.
+
+The type of the value provided by the context is inferred from the value passed to the `createContext` call:
+
+<Sandpack>
+
+```tsx App.tsx active
+import { createContext, useContext, useState } from 'react';
+
+type Theme = "light" | "dark" | "system";
+const ThemeContext = createContext<Theme>("system");
+
+const useGetTheme = () => useContext(ThemeContext);
+
+export default function MyApp() {
+  const [theme, setTheme] = useState<Theme>('light');
+
+  return (
+    <ThemeContext.Provider value={theme}>
+      <MyComponent />
+    </ThemeContext.Provider>
+  )
+}
+
+function MyComponent() {
+  const theme = useGetTheme();
+
+  return (
+    <div>
+      <p>Current theme: {theme}</p>
+    </div>
+  )
+}
+```
+
+```js App.js hidden
+import AppTSX from "./App.tsx";
+export default App = AppTSX;
+```
+
+</Sandpack>
+
+This technique works when you have an obvious default value - but often with context you do not, and [in those cases `null` is an appropriate default value](/reference/react/useContext#specifying-a-fallback-default-value). However, you likely would not want `null` to be in the type when consuming the type, our recommendation is to have the hook do a runtime check for it's existence and throw an error when not present:
```

Comment:
This example is IMO an anti-pattern if it's included it should at least be marked as such. That's why the doc that's linked is specifically talking about specifying a default value _other than_ null.

I think this can be done pretty simply by re-iterating what the linked doc says eg.

> Often, instead of null, there is some more meaningful value you can use as a default. For the occasions where this not...

Reply:
There's a very real chance this is just me - but in the current codebase I work in there are zero cases where context has a default value (there's also a lot of `{} as any` or `null as any`)   - in some of the Artsy codebases I looked at it was about half of the `createContext` had a value like those too.

I'm sure I can go back and maybe add a fake object which conforms to the same shape with NOOP funcs etc, but doing all that dancing for no useful value seems like busywork vs this technique which raises in the right location when there's an issue and acts like it is supposed to act inside components

But I agree, there's a contradiction in what this side is saying vs what the docs maybe there's a way to re-frame this paragraph 👍🏻

Reply:
In my experience it's a _very prevalent_ anti pattern 😅 if it counts for anything pretty sure I've seen React core team members say similar things about context before, but I don't want to derail this discussion too much about context idioms 😆 

It's a pattern common enough to be documented here, just needs to match the tone of "this is probably not a good thing to be doing" IMO.

Reply:
Sounds good!